cli_tools/common/utils/param/network_resolver.go

```diff
@@ -0,0 +1,127 @@
+//  Copyright 2021 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package param
+
+import (
+	"strings"
+
+	"google.golang.org/api/compute/v1"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/paramhelper"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+)
+
+// NetworkResolver standardizes and validates network and subnet fields. It follows the
+// rules from the `networkInterfaces[].network` section of instances.insert:
+//    https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert
+type NetworkResolver interface {
+	// Resolve returns the URI representation of network and subnet
+	// within a given region.
+	//
+	// There are two goals:
+	//
+	//    a. Explicitly use the 'default' network only when
+	//       network is omitted and subnet is empty.
+	//    b. Convert bare identifiers to URIs.
+	Resolve(originalNetwork, originalSubnet, region, project string) (network, subnet string, err error)
+}
+
+// NewNetworkResolver returns a NetworkResolver implementation that uses the Compute API.
+func NewNetworkResolver(client daisyCompute.Client) NetworkResolver {
+	return &computeNetworkResolver{client}
+}
+
+// computeNetworkResolver uses the Compute API to implement NetworkResolver.
+type computeNetworkResolver struct {
+	client daisyCompute.Client
+}
+
+func (r *computeNetworkResolver) Resolve(
+	originalNetwork, originalSubnet, region, project string) (network, subnet string, err error) {
+
+	// 1. Segment the user's input into component fields such as network name, subnet name, project, and region.
+	// If the URI in originalNetwork or originalSubnet didn't specify project or region, then backfill
+	// those fields using region and project.
+	networkFields, subnetFields, err := parseNetworkAndSubnet(originalNetwork, originalSubnet, region, project)
```

Comment:
networkFields and subnetFields are of types NetworkResource and SubnetResource. Having var fields match type names would make it easier to follow the code.

Reply:
updated